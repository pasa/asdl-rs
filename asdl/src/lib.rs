mod model;
mod parser;

use std::error::Error;
use std::fmt;

pub use model::Asdl;

pub type Result<T> = std::result::Result<T, AsdlError>;

#[derive(Debug)]
pub struct AsdlError {
    details: String,
}

impl fmt::Display for AsdlError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        write!(f, "{}", self.details)
    }
}

impl Error for AsdlError {
    fn description(&self) -> &str {
        &self.details
    }
}

pub fn model(asdl: &str) -> Result<Asdl> {
    let (_, root) = parser::parse(asdl).unwrap();
    Ok(Asdl::new(&root))
}

#[cfg(test)]
mod tests {
    use crate::*;
    use insta::assert_debug_snapshot_matches;

    #[test]
    fn simple_successful_test() {
        let asdl = r"
            stm = Compound(stm s1, stm* s2)
                | Single(stm)
            noFileds = One | Two | Three
            prodType = (noFileds? f, stm s1)
            ";
        let (_, root) = parser::parse(&asdl).unwrap();
        assert_debug_snapshot_matches!("simple_successful_test_syntax", root);
        let model = Asdl::new(&root);
        assert_debug_snapshot_matches!("simple_successful_test_model", model)
    }

    #[test]
    fn attributes() {
        let asdl = r"
            stm = Compound(stm s1, stm* s2)
                | Single(stm)
                  attributes(prodType?)
            prodType = (stm s1)
            ";
        let (_, root) = parser::parse(&asdl).unwrap();
        assert_debug_snapshot_matches!("attributes_syntax", root);
        let model = Asdl::new(&root);
        assert_debug_snapshot_matches!("attributes_model", model)
    }
}
