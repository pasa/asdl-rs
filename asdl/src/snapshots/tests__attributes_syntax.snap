---
created: "2019-08-28T14:49:57.958175708Z"
creator: insta@0.8.1
source: asdl/src/lib.rs
expression: root
---
Root {
    types: [
        SumType(
            SumType {
                type_id: TypeId(
                    "stm",
                ),
                constructors: [
                    Constr {
                        id: ConstrId(
                            "Compound",
                        ),
                        fields: [
                            Required(
                                Required {
                                    type_id: TypeId(
                                        "stm",
                                    ),
                                    id: Some(
                                        Id(
                                            "s1",
                                        ),
                                    ),
                                },
                            ),
                            Repeated(
                                Repeated {
                                    type_id: TypeId(
                                        "stm",
                                    ),
                                    id: Some(
                                        Id(
                                            "s2",
                                        ),
                                    ),
                                },
                            ),
                        ],
                        comments: [],
                    },
                    Constr {
                        id: ConstrId(
                            "Single",
                        ),
                        fields: [
                            Required(
                                Required {
                                    type_id: TypeId(
                                        "stm",
                                    ),
                                    id: None,
                                },
                            ),
                        ],
                        comments: [],
                    },
                ],
                attrs: Some(
                    Attrs {
                        fields: [
                            Optional(
                                Optional {
                                    type_id: TypeId(
                                        "prodType",
                                    ),
                                    id: None,
                                },
                            ),
                        ],
                    },
                ),
                comments: [],
            },
        ),
        ProdType(
            ProdType {
                type_id: TypeId(
                    "prodType",
                ),
                fields: [
                    Required(
                        Required {
                            type_id: TypeId(
                                "stm",
                            ),
                            id: Some(
                                Id(
                                    "s1",
                                ),
                            ),
                        },
                    ),
                ],
                comments: [],
            },
        ),
    ],
    comments: [],
}
